### PR TITLE
Support retrieving from multiple sources; add deeplake TQL support

### PR DESCRIPTION
**Pull request by jerpint**

### CODE REVIEW

File: buster/retriever/deeplake.py [hunk 1]
```diff
@@ -9,17 +11,67 @@
...
+    embedding_string = ",".join([str(item) for item in embedding])
+
+    # If sources is provided and it's not empty, build the where clause.
+    if sources:
+        conditions = [f"contains(metadata['source'], '{source}')" for source in sources]
+        where_clause = "where " + " or ".join(conditions)
+
+    # Construct the entire query
+    query = f"""
+select * from (
```
Comment (hbertrand) [inline buster/retriever/deeplake.py:51]: I don't understand why you need the `select * from`

File: buster/retriever/deeplake.py [hunk 2]
```diff
@@ -9,17 +11,67 @@
...
+    def __init__(self, path, exec_option: str = "python", org_id: str = None, use_tql: bool = False, **kwargs):
         from deeplake.core.vectorstore import VectorStore
 
         super().__init__(**kwargs)
+        self.use_tql = use_tql
+        self.exec_option = exec_option
         self.vector_store = VectorStore(
             path=path,
             read_only=True,
+            token=os.getenv("ACTIVELOOP_TOKEN"),
```
Comment (hbertrand) [inline buster/retriever/deeplake.py:70]: should be exposed in the init and loaded elsewhere

File: buster/retriever/deeplake.py [hunk 3]
```diff
@@ -9,17 +11,67 @@
...
+            org_id=org_id,
         )
 
-    def get_documents(self, source: str = None):
+    def get_documents(self, sources: list[str] = None):
```
Comment (hbertrand) [inline buster/retriever/deeplake.py:74]: need `Optional`

File: buster/retriever/deeplake.py [hunk 4]
```diff
@@ -55,36 +107,27 @@ def get_topk_documents(
         else:
             raise ValueError("must provide either a query or an embedding")
 
-        if source is not None:
-            logger.info(f"Applying source {source} filter...")
-            filter = {"metadata": {"source": source}}
+        if self.use_tql:
+            assert self.exec_option == "compute_engine", "cant use tql without compute_engine"
+            tql_query = build_tql_query(query_embedding, sources=sources, top_k=top_k)
+            data = self.vector_store.search(query=tql_query)
```
Comment (hbertrand) [inline buster/retriever/deeplake.py:113]: what is the advantage of tql if there is already a search function implemented ?

File: buster/retriever/service.py [hunk 5]
```diff
@@ -66,14 +67,16 @@ def get_source_display_name(self, source: str) -> str:
...
+    def get_topk_documents(self, query: str, sources: Optional[list[str]], top_k: int) -> pd.DataFrame:
+        if sources is None:
             filter = None
         else:
-            filter = {"source": {"$eq": source}}
-            source_exists = self.db.sources.find_one({"name": source})
+            if len(sources) > 1:
+                raise ValueError("Service retriever only supports one source for now...")
+            filter = {"source": {"$eq": sources[0]}}
+            source_exists = self.db.sources.find_one({"name": sources[0]})
```
Comment (hbertrand) [inline buster/retriever/service.py:77]: might as well include support for this here. This should work:
```python
filter = {"source": {"$in": sources}}
source_exists = self.db.sources.find_one({"name": {"$in": sources}})
```

**Review by hbertrand — CHANGES_REQUESTED** [review]

File: buster/retriever/deeplake.py (hunk 1 above)
Comment (jerpint) [reply buster/retriever/deeplake.py:51]: not sure, it's deeplake's TQL which im not too familiar with

File: buster/retriever/deeplake.py (hunk 4 above)
Comment (jerpint) [reply buster/retriever/deeplake.py]: not sure i understand the question - search is implemented using the TQL query directly - they support search with python functions but its extremely slow

File: buster/retriever/deeplake.py (hunk 2 above)
Comment (jerpint) [reply buster/retriever/deeplake.py]: done, added a warning if the variable isn't set (you should still be able to use the library just fine without it if not using enterprise features)

File: buster/retriever/deeplake.py (hunk 4 above)
Comment (hbertrand) [reply buster/retriever/deeplake.py]: then is there any reason we need the python function version? We should always use the tql right?

Comment (jerpint) [reply buster/retriever/deeplake.py]: TQL is only supported for enterprise users so we stil lneed python for open-source free version

File: buster/retriever/deeplake.py [hunk 6]
```diff
@@ -58,20 +58,29 @@ def build_tql_query(embedding, sources=None, top_k: int = 3):
 
 
 class DeepLakeRetriever(Retriever):
-    def __init__(self, path, exec_option: str = "python", org_id: str = None, use_tql: bool = False, **kwargs):
+    def __init__(self, path, exec_option: str = "python", use_tql: bool = False, deep_memory=False, **kwargs):
         from deeplake.core.vectorstore import VectorStore
 
         super().__init__(**kwargs)
+        activeloop_token = os.getenv("ACTIVELOOP_TOKEN")
```
Comment (hbertrand) [inline buster/retriever/deeplake.py:65]: Sorry I wasn't clear, I meant having the envvar as an argument of the init. This is not great having an env var being loaded deep within a library, it should be exposed as much as possible

File: buster/retriever/deeplake.py (hunk 1 above)
Comment (jerpint) [reply buster/retriever/deeplake.py:51]: This one in particular I wrote myself based on examples from their docs, https://docs.deeplake.ai/en/latest/Tensor-Query-Language.html

Comment (hbertrand) [reply buster/retriever/deeplake.py:51]: then try this one:
`select embedding, text, metadata, cosine_similarity(embedding, ARRAY[{embedding_string}]) as score {where_clause} order by score desc limit {top_k}`

Comment (jerpint) [reply buster/retriever/deeplake.py:51]: it fails :(

Comment (hbertrand) [reply buster/retriever/deeplake.py:51]: alright then, let's forget it. It's a very weird looking query, but I don't want to look into their doc to make it nicer.

**Review by hbertrand — APPROVED** [review]

**Pull request merged**